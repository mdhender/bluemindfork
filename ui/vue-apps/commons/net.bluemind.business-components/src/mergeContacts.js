import get from "lodash.get";
import isEqual from "lodash.isequal";
import { inject } from "@bluemind/inject";

export default async function merge(contacts, contactContainerUids) {
    if (contacts?.length !== contactContainerUids?.length) {
        throw "'contacts' and 'contactContainerUids' must have the same length";
    }

    const contactsWithContainerUid = contacts.map((c, index) => ({ ...c, containerUid: contactContainerUids[index] }));
    const sortedContacts = await sortContacts(contactsWithContainerUid);
    const bestContactForIdentification = bestContact(sortedContacts, "value.identification.formatedName.value");
    const photoBinary = await fetchPhoto(bestContact(sortedContacts, "value.identification.photo"));
    return {
        value: {
            identification: {
                ...bestContactForIdentification.value?.identification,
                photo: !!photoBinary,
                photoBinary
            },
            communications: {
                tels: mergeList(sortedContacts, "value.communications.tels"),
                emails: mergeList(sortedContacts, "value.communications.emails", "value")
            },
            deliveryAddressing: mergeList(sortedContacts, "value.deliveryAddressing")
        },
        uid: bestContactForIdentification.uid,
        containerUid: bestContactForIdentification.containerUid
    };
}

function bestContact(sortedContacts, propPath) {
    return sortedContacts.find(c => !!get(c, propPath, c));
}

function mergeList(sortedContacts, listPath, propPath) {
    const items = [];
    sortedContacts.forEach(c =>
        get(c, listPath)?.forEach(item => {
            const itemValue = get(item, propPath, item);
            if (!items.some(i => isEqual(get(i, propPath, i), itemValue))) {
                items.push(item);
            }
        })
    );
    return items;
}

async function sortContacts(contacts) {
    const { userId, domain } = inject("UserSession");
    return [...contacts].sort((a, b) => {
        return priority(b.containerUid, userId, domain) - priority(a.containerUid, userId, domain);
    });
}

function priority(containerUid, userId, domain) {
    if (containerUid === `addressbook_${domain}`) {
        return 3;
    }
    if (containerUid === `book:Contacts_${userId}`) {
        return 2;
    }
    if (containerUid === `book:CollectedContacts_${userId}`) {
        return 0;
    }
    return 1;
}

async function fetchPhoto(contact) {
    if (contact) {
        const photoURL = `api/addressbooks/${contact.containerUid}/${contact.uid}/photo`;
        return await fetch(photoURL);
    }
}
