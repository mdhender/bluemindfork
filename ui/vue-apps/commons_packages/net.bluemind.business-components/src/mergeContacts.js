import get from "lodash.get";
import isEqual from "lodash.isequal";
import { isCollectAddressBook, isDirectoryAddressBook, isPersonalAddressBook } from "@bluemind/contact";
import { inject } from "@bluemind/inject";

export default async function merge(contactsWithContainerUid) {
    const sortedContacts = await sortContacts(contactsWithContainerUid);
    const bestContactForIdentification = bestContact(sortedContacts, "value.identification.formatedName.value");
    const photoBinary = await photoUrl(bestContact(sortedContacts, "value.identification.photo"));
    const defaultEmailFn = email =>
        email.parameters.find(({ label, value }) => label === "DEFAULT" && value === "true") ? 1 : 0;
    const sortEmailsFn = (a, b) => defaultEmailFn(b) - defaultEmailFn(a);
    return {
        value: {
            identification: {
                ...bestContactForIdentification.value?.identification,
                photo: !!photoBinary,
                photoBinary
            },
            communications: {
                tels: mergeList(sortedContacts, "value.communications.tels"),
                emails: mergeList(sortedContacts, "value.communications.emails", "value").sort(sortEmailsFn)
            },
            deliveryAddressing: mergeList(sortedContacts, "value.deliveryAddressing")
        },
        uid: bestContactForIdentification.uid,
        containerUid: bestContactForIdentification.containerUid
    };
}

function bestContact(sortedContacts, propPath) {
    return sortedContacts.find(c => !!get(c, propPath, c));
}

function mergeList(sortedContacts, listPath, propPath) {
    const items = [];
    sortedContacts.forEach(c =>
        get(c, listPath)?.forEach(item => {
            const itemValue = get(item, propPath, item);
            if (!items.some(i => isEqual(get(i, propPath, i), itemValue))) {
                items.push(item);
            }
        })
    );
    return items;
}

async function sortContacts(contacts) {
    const { userId, domain } = inject("UserSession");
    return [...contacts].sort((a, b) => {
        return priority(b.containerUid, userId, domain) - priority(a.containerUid, userId, domain);
    });
}

function priority(containerUid, userId, domain) {
    if (isDirectoryAddressBook(containerUid, domain)) {
        return 3;
    }
    if (isPersonalAddressBook(containerUid, userId)) {
        return 2;
    }
    if (isCollectAddressBook(containerUid, userId)) {
        return 0;
    }
    return 1;
}

async function photoUrl(contact) {
    if (contact) {
        return `${window.location.protocol}//${window.location.hostname}/api/addressbooks/${contact.containerUid}/${contact.uid}/photo`;
    }
}
