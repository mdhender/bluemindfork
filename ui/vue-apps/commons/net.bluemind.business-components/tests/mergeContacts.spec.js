import mergeContacts from "../src/mergeContacts";

jest.mock("@bluemind/inject", () => {
    const getCompleteBookResults = require("./data/getCompleteBookResults.json");
    return {
        inject: service => {
            if (service === "AddressBooksMgmtPersistence") {
                return { getComplete: uid => getCompleteBookResults[uid] };
            }
            if (service === "UserSession") {
                return { userId: "", domain: "" };
            }
        }
    };
});

global.fetch = jest.fn(() => "photoBinary");

describe("Merge contacts", () => {
    test("Merge contacts from different address books", async () => {
        const searchResults = require("./data/searchResults.json");
        const getCompleteResults = require("./data/getCompleteResults.json");
        const contact = await mergeContacts(
            getCompleteResults,
            searchResults.values.map(v => v.containerUid)
        );
        expect(contact).toMatchSnapshot();
    });
});
