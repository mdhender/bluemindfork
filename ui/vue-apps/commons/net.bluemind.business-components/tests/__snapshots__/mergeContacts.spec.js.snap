// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Merge contacts Merge contacts from different address books 1`] = `
Object {
  "containerUid": "addressbook_75a0d5b3.internal",
  "uid": "E34DD2DD-25AC-4548-927F-C7D183203608",
  "value": Object {
    "communications": Object {
      "emails": Array [
        Object {
          "parameters": Array [
            Object {
              "label": "TYPE",
              "value": "home",
            },
          ],
          "value": "georges.abitbol@dev.bluemind.net",
        },
        Object {
          "parameters": Array [
            Object {
              "label": "TYPE",
              "value": "work",
            },
          ],
          "value": "gabitbol@devenv.dev.bluemind.net",
        },
        Object {
          "parameters": Array [
            Object {
              "label": "TYPE",
              "value": "work",
            },
          ],
          "value": "georges.a@dev.bluemind.net",
        },
        Object {
          "parameters": Array [
            Object {
              "label": "TYPE",
              "value": "home",
            },
          ],
          "value": "georges.bitbol@dev.bluemind.net",
        },
        Object {
          "parameters": Array [
            Object {
              "label": "TYPE",
              "value": "work",
            },
          ],
          "value": "georges.b@dev.bluemind.net",
        },
      ],
      "tels": Array [
        Object {
          "ext": null,
          "parameters": Array [
            Object {
              "label": "TYPE",
              "value": "voice",
            },
            Object {
              "label": "TYPE",
              "value": "work",
            },
          ],
          "value": "0615249193",
        },
        Object {
          "ext": null,
          "parameters": Array [
            Object {
              "label": "TYPE",
              "value": "home",
            },
            Object {
              "label": "TYPE",
              "value": "voice",
            },
          ],
          "value": "0615249194",
        },
        Object {
          "ext": null,
          "parameters": Array [
            Object {
              "label": "TYPE",
              "value": "voice",
            },
            Object {
              "label": "TYPE",
              "value": "work",
            },
          ],
          "value": "0615249100",
        },
        Object {
          "ext": null,
          "parameters": Array [
            Object {
              "label": "TYPE",
              "value": "home",
            },
            Object {
              "label": "TYPE",
              "value": "voice",
            },
          ],
          "value": "0615249101",
        },
      ],
    },
    "deliveryAddressing": Array [
      Object {
        "address": Object {
          "countryName": "France",
          "extentedAddress": null,
          "locality": "Tolosa",
          "parameters": Array [
            Object {
              "label": "TYPE",
              "value": "work",
            },
          ],
          "postOfficeBox": "BP 007",
          "postalCode": "31000",
          "region": "",
          "streetAddress": "6 avenue du Pont",
          "value": null,
        },
      },
    ],
    "identification": Object {
      "anniversary": null,
      "birthday": null,
      "formatedName": Object {
        "parameters": Array [],
        "value": "Georges A",
      },
      "gender": Object {
        "parameters": Array [],
        "text": null,
        "value": null,
      },
      "name": Object {
        "additionalNames": "",
        "familyNames": "A",
        "givenNames": "Georges",
        "parameters": Array [],
        "prefixes": "",
        "suffixes": "",
        "value": null,
      },
      "nickname": Object {
        "parameters": Array [],
        "value": null,
      },
      "photo": true,
      "photoBinary": "photoBinary",
    },
  },
}
`;
